File: Add Code Here/C++/leetcode/4Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& arr, int k) {
        int n = arr.size();
        sort(arr.begin(), arr.end());
        vector<vector<int>>v;
        for(int i=0;i<n-3;) {
            for(int j=i+1;j<n-2;) {
                int sum = k - (arr[i]+arr[j]);
                int l = j+1, r=n-1;
                while(l<r) {
                    if(sum==arr[l]+arr[r])
                    {
                        v.push_back({arr[i],arr[j],arr[l],arr[r]});
                        int t = arr[l], fr = arr[r];
                        while(l<r && arr[l]==t)
                            l++;
                        while(l<r && arr[r]==fr) r--;
                    }
                    else if(sum<arr[l]+arr[r])
                        r--;
                    else l++;
                }
                int s = arr[j];
                while(j<n-2 && arr[j]==s) j++;
            }
            int f = arr[i];
            while(i<n-3 && arr[i]==f) i++;
        }
        return v;
    }
};

File: Add Code Here/JAVA/4Sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        Arrays.sort(nums);
        Set<List<Integer>>set = new HashSet();
        for(int i=0;i<n;i++){
            
            for(int j=i+1;j<n;j++){
                
                int k = j+1;
                int l = n-1;
                while(k<l){
                    long sum = nums[i]+nums[j];
                    sum += nums[k];
                    sum += nums[l];
                if(sum==target){
                    List<Integer>arr = new ArrayList();
                    arr.add(nums[i]);
                    arr.add(nums[j]);
                    arr.add(nums[k]);
                    arr.add(nums[l]);
                    set.add(arr);
                    k++;
                    l--;
                    while(k<l && nums[k] == nums[k-1])k++;
                    while(k<l && nums[l] == nums[l+1])l--;
                }
                else if(sum<target)k++;
                else l--;
                }
                
            }
        }
        return new ArrayList(set);
    }
}